src/BSON/UTCDateTime.c

```diff
@@ -61,8 +61,42 @@ static bool php_phongo_utcdatetime_init(php_phongo_utcdatetime_t *intern, int64_
 	return true;
 }
 
-/* Initialize the object from a HashTable and return whether it was successful. */
-static bool php_phongo_utcdatetime_init_from_hash(php_phongo_utcdatetime_t *intern, HashTable *props)
+/* Initialize the object from a numeric string and return whether it was
+ * successful. An exception will be thrown on error. */
+static bool php_phongo_utcdatetime_init_from_string(php_phongo_utcdatetime_t *intern, const char *s_milliseconds, phongo_zpp_char_len s_milliseconds_len TSRMLS_DC)
+{
+	int64_t milliseconds;
+	char *endptr = NULL;
+
+	errno = 0;
+
+#if defined(PHP_WIN32)
+	milliseconds = _atoi64(s);
+#else
+	milliseconds = bson_ascii_strtoll(s_milliseconds, &endptr, 10);
+#endif
+
+	/* errno will set errno if conversion fails; however, we do not need to
+	 * specify the type of error.
+	 *
+	 * Note: bson_ascii_strtoll() does not properly detect out-of-range values
+	 * (see: CDRIVER-1377). strtoll() would be preferable, but it is not
+	 * available on all platforms (e.g. HP-UX), and atoll() provides no error
+	 * reporting at all. */
```

Comment:
The original `STRTOLL()` macro we had uses three possible functions:

 * `_atoi64()` on Windows, which supports error reporting (may not detect extra data)
 * `atoll()` if `HAVE_ATOLL` is defined, which doesn't support error reporting at all
 * `strtoll()` otherwise, which supports error reporting and `endptr` for detecting extra data

Support for `atoll()` appears to be a concession for HP-UX's broken `strtoll()` implementation (see: [PHP bug #33433](https://bugs.php.net/bug.php?id=33433)). If we care about HP-UX portability, I suggest we use `bson_ascii_strtoll()`, which is a `strtoll()` polyfill in libbson. It does have an issue with detecting out-of-range integers (see: [CDRIVER-1377](https://jira.mongodb.org/browse/CDRIVER-1377)), but I think this is an improvement over the current use of `atoll()` and reporting no errors at all.